### PR TITLE
Vive Pro Defaults OpenVr Generic Device causing one hand operation errors due to missing Spatial Grip Input Action

### PR DESCRIPTION
**Pull request by wiwei**
https://github.com/microsoft/MixedRealityToolkit-Unity/issues/4470

Adds the Input Action assignment to support Vive Pro controllers

### CODE REVIEW

**Comment by wiwei** [comment]
For this change, I applied the fix that was mentioned on slack (and linked in the issue).

Unfortunately I don't have access to Vive Pro hardware - @JShull can you confirm that this addresses the issue that you hit?

**Comment by JShull** [comment]
The Vive Pro defaults to the generic openVR controller but the spatial grip reference wasn't on the generic controller mapping so it throws an error the first time you go to one hand use the controller - not sure why it's looking for the spatial grip when I am using the trigger but it does! When I added in the spatial grip to the generic openVR controller mapping the error went away.

**Comment by JShull** [comment]
Confirmed working with Vive Pro Hardware, the error no longer comes up and one hand pick up and move works as it should.

File: Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset [hunk 1]
```diff
@@ -2870,6 +2870,19 @@ MonoBehaviour:
       axisCodeY: AXIS_18
       invertXAxis: 0
       invertYAxis: 0
+    - id: 11
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset]: We should probably make sure the ID, description, and any other stale data from the create button (axisCodeX/Y as well) are reset for this action.

Might be an unrelated bug in the generic action mapper that everything is copied from the previous action, even if you aren't able to change it in the UI.

File: Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset [hunk 2]
```diff
@@ -2870,6 +2870,19 @@ MonoBehaviour:
...
+      description: WMR Touchpad Position
+      axisType: 7
+      inputType: 14
+      inputAction:
+        id: 3
+        description: Grip Pose
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset]: These should be added to the default interactions in the class as well:

https://github.com/microsoft/MixedRealityToolkit-Unity/blob/9ab0e9b69afe45f7856b5731436ae427c3fe7cc4/Assets/MixedRealityToolkit.Providers/OpenVR/GenericOpenVRController.cs#L37

**Comment by luis-valverde-ms** [comment]
@wiwei , is this still alive?

File: Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset (hunk 1 above)
Comment (wiwei) [reply Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset]: Perfect, thanks!

---
In reply to: [286192378](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4484#discussion_r286192378) [](ancestors = 286192378)

**Comment by wiwei** [comment]
Yep, sorry for the delay here. I got pulled into doing a bunch of other random work, but finally got back to this.

**Comment by SimonDarksideJ** [comment]
Since you have made changes to the generic controller, have you confirmed this has no knock on impacts to other controllers?  I would have expected Vive Wand controller updates in the ViveWand controller class
https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk_release/Assets/MixedRealityToolkit.Providers/OpenVR/ViveWandController.cs

Likely if the Vive Pro is reporting it's controllers differently to the base vive, in which case, it should have it's own definition anyway,

File: Assets/MixedRealityToolkit.Providers/OpenVR/GenericOpenVRController.cs [hunk 3]
```diff
@@ -82,6 +82,7 @@ public GenericOpenVRController(TrackingState trackingState, Handedness controlle
             new MixedRealityInteractionMapping(9, "Unity Button Id 3", AxisType.Digital, DeviceInputType.ButtonPress, KeyCode.JoystickButton3),
             new MixedRealityInteractionMapping(10, "WMR Touchpad Touch", AxisType.Digital, DeviceInputType.TouchpadTouch, KeyCode.JoystickButton18),
             new MixedRealityInteractionMapping(11, "WMR Touchpad Position", AxisType.DualAxis, DeviceInputType.Touchpad, ControllerMappingLibrary.AXIS_17, ControllerMappingLibrary.AXIS_18),
+            new MixedRealityInteractionMapping(12, "WMR Grip Pose", AxisType.SixDof, DeviceInputType.SpatialGrip, MixedRealityInputAction.None),
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.Providers/OpenVR/GenericOpenVRController.cs]: Is this specific to WMR? I believe this should work regardless of controller / platform: https://github.com/microsoft/MixedRealityToolkit-Unity/blob/f3a97916629a965c9fc45e85290c453ebc9b45f3/Assets/MixedRealityToolkit/Providers/UnityInput/GenericJoystickController.cs#L207

Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit.Providers/OpenVR/GenericOpenVRController.cs]: Arguably, it's a mistake when I first put that forward as we only had WMR to test on.  These should all be called "Generic" as they are per the unaltered generic spec, not specific to any device.

**Comment by keveleigh** [comment]
>Since you have made changes to the generic controller, have you confirmed this has no knock on impacts to other controllers? I would have expected Vive Wand controller updates in the ViveWand controller class

Yeah, it's possible this mapping is missing from other OpenVR controllers as well (it does appear to be missing from the Vive Wand class). We should probably log this or go through in a subsequent PR to ensure feature parity across classes that support them.

>Likely if the Vive Pro is reporting it's controllers differently to the base vive, in which case, it should have it's own definition anyway

That makes sense, but this PR is a good first step in that direction by at least unblocking devs from using that controller with the Toolkit.

**Comment by luis-valverde-ms** [comment]
@wiwei , are you awaiting reviews or have pending changes to implement?

**Comment by wiwei** [comment]
Thanks for the reminder :)

**Pull request merged**